pkg/serviceaccount/jwt_external.go

```diff
@@ -0,0 +1,174 @@
+
+	return &externalJWTTokenGenerator{
+		iss:    iss,
+		client: externalsigner.NewJWTKeyServiceClient(conn),
+	}, nil
+}
+
+type externalJWTTokenGenerator struct {
+	iss    string
+	client externalsigner.JWTKeyServiceClient
+}
+
+func (g *externalJWTTokenGenerator) GenerateToken(claims *jwt.Claims, privateClaims interface{}) (string, error) {
+
+	signer := NewRemoteOpaqueSigner(g.client)
+
+	generator, err := JWTTokenGenerator(g.iss, signer.Public().KeyID, signer)
+	if err != nil {
+		return "", err
+	}
+	return generator.GenerateToken(claims, privateClaims)
+}
+
+// NewRemoteOpaqueSigner returns an jose.OpaqueSigner that communicates over the client
+func NewRemoteOpaqueSigner(client externalsigner.JWTKeyServiceClient) jose.OpaqueSigner {
+	return &opaqueSigner{client: client}
+}
+
+type opaqueSigner struct {
+	client externalsigner.JWTKeyServiceClient
+}
+
+// check that opaqueSigner conforms to the interface
+var _ jose.OpaqueSigner = &opaqueSigner{}
+
+func (s *opaqueSigner) Public() *jose.JSONWebKey {
+	resp, err := s.client.GetPublicKey(context.Background(), &externalsigner.GetPublicKeyRequest{})
+	if err != nil {
+		klog.Errorf("Error getting public key %v", err)
+		return nil
+	}
+	response := &jose.JSONWebKey{
+		KeyID:     resp.PublicKey.KeyId,
+		Algorithm: resp.PublicKey.Algorithm,
+		Use:       resp.PublicKey.Use,
+	}
+
+	keys, err := certutil.ParsePublicKeysPEM(resp.PublicKey.PublicKey)
+	if err != nil {
+		klog.Errorf("Error getting public key %v", err)
+		return nil
+	}
+	if len(keys) == 0 {
+		klog.Error("No public key returned")
+		return nil
+	}
+	response.Key = keys[0]
+
+	response.Certificates, err = certutil.ParseEmptyCertsPEM(resp.PublicKey.Certificates)
+	if err != nil {
+		klog.Errorf("Error parsing x509 certificate %v", err)
+		return nil
+	}
+	return response
+}
+
+func (s *opaqueSigner) Algs() []jose.SignatureAlgorithm {
+	response := []jose.SignatureAlgorithm{}
+	algsResp, err := s.client.ListAlgorithms(context.Background(), &externalsigner.ListAlgorithmsRequest{})
+	if err != nil {
+		klog.Errorf("Error listing algorithms %v", err)
+		return response
+	}
+	for _, alg := range algsResp.SignatureAlgorithms {
+		response = append(response, jose.SignatureAlgorithm(alg))
+	}
+	return response
+}
+
+func (s *opaqueSigner) SignPayload(payload []byte, alg jose.SignatureAlgorithm) ([]byte, error) {
+	resp, err := s.client.SignPayload(context.Background(), &externalsigner.SignPayloadRequest{
+		Payload:   payload,
+		Algorithm: string(alg),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.Content, nil
+}
+
+type externalJwtTokenAuthenticator struct {
+	iss          string
+	client       externalsigner.JWTKeyServiceClient
+	validator    Validator
+	implicitAuds authenticator.Audiences
+}
+
+func (a *externalJwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData string) (*authenticator.Response, bool, error) {
+
+	keyResp, err := a.client.ListPublicKeys(ctx, &externalsigner.ListPublicKeysRequest{})
```

Comment:
It seems expensive to re-fetch public keys to validate every token

Reply:
Is it worth adding a cache?